File: Python/sorting/merge_sort.py
# Python program for implementation of MergeSort
 
# merge sort works by dividing an array into two halves, sorting each of 
# those halves, and then putting the array back together. This process is repeated
# until the entire ray has been sorted successfully. 

# merge_sort main
def merge_sort(arr):
    if len(arr) < 2:
        print ("Array is already sorted")
        return arr
    
    # Split the array into two halves
    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]
    
    # Recursively sort each half
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)
    
    # Merge the sorted halves
    return merge(left_half, right_half)


# merge function for knitting the arrays back together
def merge(left, right):
    merged = []
    i = j = 0
    
    # put the two arrays back together
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    
    # Append remaining elements from left and right arrays
    while i < len(left):
        merged.append(left[i])
        i += 1
    while j < len(right):
        merged.append(right[j])
        j += 1
    
    return merged

# Sample array example
arr = [5, 2, 8, 4, 1, 9, 3, 6, 7]
print(arr)
sorted_arr = merge_sort(arr)
print("Sorted array:", sorted_arr)

File: Java/sorting/MergeSort.java
// merge sort implementation in Java. 

// This algorithm works by recursively dividing an array, 
// sorting each half, and then putting them back together

// necessary imports
import java.util.Arrays;

public class MergeSort {
    public static void mergeSort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return; 
        }
        int n = arr.length;
        int[] temp = new int[n]; 

        // main merge sort operations
        mergeSortMain(arr, 0, n - 1, temp);
    }

    // finds the middle of the array and uses recursion to sort the right and left halves
    private static void mergeSortMain(int[] arr, int low, int high, int[] temp) {
        if (low < high) {
            int mid = low + (high - low) / 2; 
            mergeSortMain(arr, low, mid, temp);
            mergeSortMain(arr, mid + 1, high, temp);

            // Merge the sorted halves
            merge(arr, low, mid, high, temp);
        }
    }

    // function responsible for merging the pieces back together 
    private static void merge(int[] arr, int lower, int middle, int high, int[] temp) {
        int i = lower; 
        int j = middle + 1; 
        int k = lower; 

        // Merge two subarrays
        while (i <= middle && j <= high) {
            if (arr[i] <= arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }

        // Copy left subarray
        while (i <= middle) {
            temp[k++] = arr[i++];
        }

        // Copy right subarray
        while (j <= high) {
            temp[k++] = arr[j++];
        }

        // Copy elements back to original array
        for (k = lower; k <= high; k++) {
            arr[k] = temp[k];
        }
    }

    // sample situation. Sorts array arr
    public static void main(String[] args) {
        int[] arr = {10, 7, 8, 9, 1, 5};
        System.out.println("Original array: " + Arrays.toString(arr));
        mergeSort(arr);
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}
